res/controllers/Novation-Launchpad MK2-scripts.js

```diff
@@ -1,5 +1,4 @@
 var NovationLaunchpadMK2 = (function () {
-'use strict';
```

Comment:
what's the problem with 'use strict'?

Reply:
@Be-ing

Reply:
I'm not sure why it wasn't working with `use strict`, but removing it got the test to pass :man_shrugging:

Reply:
Can you test with real hardware to confirm it actually works beyond passing the automated test?

Reply:
I'm not sure `use strict` actually did anything in the old QtScript interpreter. `use strict` was [added in ES5](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Strict_mode#Specifications).

Reply:
the script does not work for me actually. It fails deep inside an internal lodash function for equality checking.

Reply:
it somehow thinks the `undefined` is an array and invoke `byteLength` on it.

Reply:
doesn't help

Reply:
:/ Well then, the other option is putting `use strict` back and figuring out why that doesn't work...

Reply:
unlikely this has anything to do with 'use strict'

Reply:
Right, there was a different problem with `use strict`. Judging from my [commit log](ecedd59ff81037221a09514a63de9c09e42b19be) I think the error only happened when building this branch with Qt 5.12. What version of Qt are you using?

Reply:
5.11.1

Reply:
If you put `use strict` back, does the script still not work with Qt 5.11?

Reply:
there's something horribly wrong here. `switch` does not work correctly. the problem is on line `6001`
although we fall into `case arrayBufferTag$3` (which is `'[object ArrayBuffer]'`), `tag` is actually `'[object DataView]'`.

Reply:
ahh nvm, we didn't break out from the previous one

Reply:
I have no idea what's going wrong there and I think you're the only one familiar enough with the code to fix it. When you get it working, please make a pull request for @ferranpujolcamins' fork and he can merge it into this branch before we merge it upstream. Also please try building with Qt 5.12 to ensure the changes to the JS interpreter don't break it again. I don't know what OS you're using, but I put some notes on the [wiki](https://mixxx.org/wiki/doku.php/compiling_on_linux#non-system_qt) for how to build with a different version of Qt than a Linux distro packages. Leave a comment here when you open that pull request so people not subscribed to notifications on @ferranpujolcamins' fork are aware of the PR.

Reply:
I have no idea as well. I don't even know where these `ArrayBuffer`s or `DataView`s are coming from. I don't even use them.

Reply:
We already report the stack trace Qt gives us when there is an unhandled exception. If there is no exception thrown, you might have luck with [console.trace](http://doc.qt.io/qt-5/qtquick-debugging.html#trace).

Reply:
I get a `ReferenceError: console is not defined` if I try that. If I'm correct console extensions are not enabled by default for QJSEngine, and have to be [installed explicitly](https://doc.qt.io/qt-5/qjsengine.html#installExtensions). See also [this doc](https://doc.qt.io/qt-5/qjsengine.html#Extension-enum)

Reply:
It seems to be working for Qt 5.12

Reply:
Yes, sorry for the ambiguity. The controller scripts seem to be working. At least there is no error on the JS side, and incoming messages are captured trigger the expected effects, however but outgoing MIDI messages are not getting to the controller. I see warnings like this:

```
Warning [Controller]: Error sending short message "Launchpad MK2: outgoing: status 0x0"
Warning [Controller]: PortMidi error: PortMidi: `Invalid MIDI message Data'
Debug [Controller]: Connected ([Channel2], beatloop_1_enabled) to connection {98c2e369-7cea-4df7-a3e3-af1ba26aab59}
```

Reply:
> Oh, thanks for looking into that. If you want to edit the C++ to enable that extension, go for it. 

installing is as simple as calling 

```
m_pScriptEngine->installExtensions(QJSEngine::ConsoleExtension);
```
in `ControllerEngine::initializeScriptEngine`
However I am still trying to figure out how to grab the std output.

Reply:
Are you running Mixxx with the `--controllerDebug` command line option?

Reply:
yes, it print stacktraces, but not console messages. Btw the stacktrace output escapes the new lines so it looks bad:

```
Warning [Controller]: ControllerEngine: "Uncaught exception at line 20470 in file file:///Users/david/mixxx/res/controllers/Novation-Launchpad MK2-scripts.js: TypeError: Cannot read property '0' of undefined\nBacktrace:\n@file:///Users/david/mixxx/res/controllers/Novation-Launchpad MK2-scripts.js:20470\nupdateLayout@file:///Users/david/mixxx/res/controllers/Novation-Launchpad MK2-scripts.js:20465\nonUnmount@file:///Users/david/mixxx/res/controllers/Novation-Launchpad MK2-scripts.js:20580\nunmount@file:///Users/david/mixxx/res/controllers/Novation-Launchpad MK2-scripts.js:18628\nonUnmount@file:///Users/david/mixxx/res/controllers/Novation-Launchpad MK2-scripts.js:20719\nunmount@file:///Users/david/mixxx/res/controllers/Novation-Launchpad MK2-scripts.js:18628\nonUnmount@file:///Users/david/mixxx/res/controllers/Novation-Launchpad MK2-scripts.js:20870\nunmount@file:///Users/david/mixxx/res/controllers/Novation-Launchpad MK2-scripts.js:18628\n@file:///Users/david/mixxx/res/controllers/Novation-Launchpad MK2-scripts.js:20884"
```

Reply:
I am still getting the MIDI errors. I don't know what causes it.

Reply:
Mixxx should send messages like this:
```
Launchpad MK2: outgoing: status 0x90 (ch 1, opcode 0x9), ctrl 0x58, val 0x07
```
instead it sends
```
Launchpad MK2: outgoing: status 0x0
```

Reply:
Can you test sending some hardcoded values like `midi.sendShortMsg(0x90, 0x58, 0x07);`? If that works, then I think the problem is in a change in how your JS code is being interpreted.

Reply:
It seems string in MIDI messages are no longer working. I was previously sending `midi.sendShortMsg('0x90', '0x58', 0x07)`. Changing them to numbers solved the issue. With this, it looks like everything is working correctly. I opened a PR against this with my fixed sources.

Reply:
See https://github.com/ferranpujolcamins/mixxx/pull/10. Or should I open against master instead? Contains only non-breaking changes. I also tested against Mixxx 2.1.

Reply:
I'm surprised passing strings to `midi.sendShortMsg` worked before. AFAIK that has never been documented and I never noticed any scripts doing that.

Reply:
Correct. .sendShortMsg is only intended to take numbers. (Strings happened to work probably because of the QScriptValue() cast.)